go/src/koding/socialapi/transport.go

```diff
@@ -0,0 +1,329 @@
+package socialapi
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const maxBodyLen = 2 << 20 // 2 MiB
+
+// contextKey is a value for use with context.WithValue.
+type contextKey struct {
+	name string
+}
+
+// SessionContextKey is a context key. It can be used in
+// HTTP handlers to attach a client session to a request.
+//
+// The session is then read by the Transport and used
+// for authentication.
+//
+// The associated value will be of *Session type.
+var SessionContextKey = &contextKey{"client-session"}
+
+// DirectorContextKey is a context key. It can be used
+// in HTTP handlers to obtain an original request
+// before it was edited by a DirectorTransport.
+//
+// The associated value will be of *http.Request type.
+var DirectorContextKey = &contextKey{"original-request"}
+
+// Session represents a user session.
+type Session struct {
+	// ClientID is an ID of the session.
+	ClientID string
+
+	// Username of the user holding the session.
+	Username string
+
+	// Team, which the user belongs to.
+	Team string
+}
+
+// Key is used for SessionCache to cache sessions by keys.
+func (s *Session) Key() string {
+	return s.Team + "/" + s.Username
+}
+
+// Valid implements the stack.Validator interface.
+func (s *Session) Valid() error {
```

Comment:
do we allow empty teams and usernames? If not, could you add them to validator too?